### PR TITLE
Hoodie Client API documentation

### PR DESCRIPTION
**Pull request by NimmiW**
# [Work-in-Progress branch] [Do not merge]
- hoodie client API page (WIP)
- hoodie Account API page (WIP)

Refering to the issue at https://github.com/hoodiehq/hoodie/issues/698

### CODE REVIEW

**Comment by NimmiW** [comment]
image

How Can I revoke this automatic merge?

**Review by fozy81 — CHANGES_REQUESTED** [review]
Looks good, there's 'WIP
  <3<3<3' at the bottom of each document. Could you remove, thanks.

**Review by pmbanugo — COMMENTED** [review]
Since you're working on documentation, your commit messages should start with `docs: ..` I think.

### file: docs/api/client/hoodie.rst

#### Example

In the example section, this part of the code snippet could be modified? cc @gr2m

```
var Hoodie = require('@hoodie/client')

var hoodie = new Hoodie({
  url: 'https://myhoodieapp.com',
  PouchDB: require('pouchdb')
})
```

I use the global `hoodie` object from the client script so I don't do any of sort initialisation since I've been using Hoodie. I would suggest stating in the doc that there is a global object available if the script is referenced before script files it'll be used in and that initialisation code is optional. 

Does the hoodie client api work with Node? If so, aren't the APIs there similar to what we have for **hoodie-server**. If the hoodie-client will be used as a module in the browser, the snippet will need to change to use browser based module loaders and the initialisation code could be `var hoodie = new Hoodie({ url: 'http://localhost:4984', PouchDB: PouchDB });`. I don't know if we need to show sample snippet using different module loaders or just use ES Module. 

#### Constructor

This section is a bit confusing to me, in particular the type and description columns. Is the type a PouchDB object? I don't know if there is a type referred to as Constructor, but I know about the Constructor function. Trying to figure out that section got me confused as to what to if I need to instantiate a PouchDB object using the constructor, add some custom setting/configuration, and then pass that object as a reference to Hoodie, and what to do with a custom pouch build. Maybe we can explain this section differently and add sample snippet?

The other arguments except from url, with type set as string, aren't they object types with properties like `dbName` and `remote`? Options.dbName is set to hoodie.account.id? Why is that so?  

Also, the link in the description for store options points to account-client.  

@NimmiW My comments were to get some clarification from the maintainers and also to highlight my thoughts/observations I found from the ReadMe which if you update, would be needed to also do similar changes to the ReadMe.

Good job and Well done!

**Comment by gr2m** [comment]
> Since you're working on documentation, your commit messages should start with docs: .. I think

That is correct, but don’t worry about commit messages while working on a pull request, we clean them up before we merge.

> I use the global hoodie object from the client script so I don't do any of sort initialisation since I've been using Hoodie. I would suggest stating in the doc that there is a global object available if the script is referenced before script files it'll be used in and that initialisation code is optional.

This is a bit tricky. It is true that if you load the hoodie client directly from the server at `http://localhost:8080/hoodie/client.js`, then you don’t need to do the initialisation yourself, it’s already done and you can use the `hoodie` object.

If you want to bundle the Hoodie client into your app bundle, then you will need to initialise it yourself as shown in the example. 

I would suggest we add both these information to the documentation.

For `hoodie.account`, I would remove the section on the constructor altogether and in the example on top remove the line `var account = new Account('https://example.com/account/api')`. And in all examples, replace `account.*` with `hoodie.account.*`

Does that all make sense?

**Comment by NimmiW** [comment]
@pmbanugo okay I got what you said. but few minutes ago I did a commit without 'docs' tag. By the way I will use that tip in here after commits.

@gr2m Are removing the constructor part? in hoodie account doc? 

And as I understood so far, what I have to do is 

I would remove the section on the constructor altogether and in the example on top remove the line var account = new Account('https://example.com/account/api'). And in all examples, replace account.* with hoodie.account.*

Is that alright?

**Comment by NimmiW** [comment]
Currently I am getting this error when I am trying to push my nimmi-dev branch to the origin repo.

image

I pulled the changes from the repo to my local repo also.

What can I do for this?

By the way I see this too in this PR

image

What I did was:
I did the changes,
I commited then locally.
Then I did git pull origin master.
Then I again commited the merged changes
Then I did git push origin nimmi-dev

**Comment by gr2m** [comment]
try `git push -f origin nimmi-dev`. I think the `git pull origin master` pulled the latest changes from master and rebased your local `nimmi-dev` on it, so the history changed. You don’t need to git  pull origin master while working on your branch

**Comment by NimmiW** [comment]
@gr2m yes pulling by force worked. :)

Currently I have to make that change you made in account section and complete rest of the document

**Comment by NimmiW** [comment]
@fozy81 I changed what you said :)

**Comment by gr2m** [comment]
@NimmiW I think you still have to push your local changes? The pull request does not show any new commits yet

**Comment by NimmiW** [comment]
@gr2m https://github.com/hoodiehq/hoodie/pull/706/commits 

There are 2 new commits 

image

**Comment by gr2m** [comment]
my fault, all good

**Review by fozy81 — APPROVED** [review]

**Comment by fozy81** [comment]
@NimmiW thanks for the change :smile:

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: can you remove these two paragraphs? They 1st above is good enough

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: remove these two lines

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: replace `account` with `hoodie.account`

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: remove `Constructor` section entirely

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: in the entire document, replace `account.` with `hoodie.account.`

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: can we link to `hoodie.account.ready` within this document instead of linking to GitHub?

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: remove the constructor here, just start with `hoodie.account.validate`

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: should be code:: js

File: docs/api/index.rst
Comment (gr2m) [inline docs/api/index.rst]: can you remove the two empty lines again?

**Review by gr2m — CHANGES_REQUESTED** [review]

**Comment by NimmiW** [comment]
@gr2m Ok gregor I got the Idea, I will make those changes in my next commit. And I it really interesting to collaborate with community.

**Comment by NimmiW** [comment]
Hi all, I am again facing an issue in make html. If anyone can give me a help in that, pls leave a comment in issue https://github.com/hoodiehq/hoodie/issues/698

**Comment by NimmiW** [comment]
@gr2m I still have to do the hyperlink change you mentioned. Since I am facing an inssue in rendering the html, I could not test it properly. I will correct that soon I fix my current issue.

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: can you make just one empty line like in the other examples?

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [inline docs/api/client/hoodie.account.rst]: hoodie.account.fetch

File: docs/api/index.rst
Comment (gr2m) [inline docs/api/index.rst:11]: can you remove this empty line here?

**Comment by gr2m** [comment]
@NimmiW I know this is one of your first PRs, and probably your biggest one yet. Just want to make sure you know that there is no problem at all if it takes a lot of iterations until you got everything right, this is totally normal :) You are doing a great job 👏

File: docs/api/client/hoodie.account.rst
Comment (NimmiW) [reply docs/api/client/hoodie.account.rst]: ok @gr2m :)

File: docs/api/index.rst
Comment (NimmiW) [reply docs/api/index.rst:11]: ok @gr2m :)

**Comment by NimmiW** [comment]
@gr2m Pls excuse me for a while becoz I did not see the comment untill I do my recent push. Yes I will take my time and do the documentation. :)

**Review by fozy81 — APPROVED** [review]

**Review by fozy81 — APPROVED** [review]

**Comment by gr2m** [comment]
Hey @NimmiW what’s the status, are you still working on it or is it ready for a final review? I think the `hoodie.log.rst` and `hoodie.connection-status.rst` are still pending, right?

**Comment by NimmiW** [comment]
@gr2m I will create PR today for what I have done so far. there some areas in the readme files of Client APIs to be cleared. And completed so that I can add those missing areas to the documentation. As you mentioned `hoodie.log.rst` and `hoodie.connection-status.rst` are still pending. I did not went through them yet. Probably today I will do that :)

**Comment by NimmiW** [comment]
@gr2m  and @fozy81 can you pls have a look on my PR? :) I want it to be reviewed

File: docs/api/client/hoodie.account.rst
Comment (pmbanugo) [inline docs/api/client/hoodie.account.rst:666]: What do you think about explaining what a hook is and how it might be useful? 

I think we should provide an explanation for what a hook is, why one might need it and how it can be used (which the sample code already shows). Then stating that it's implemented with that npm package and providing the link to it

**Review by fozy81 — APPROVED** [review]

File: docs/conf.py
Comment (gr2m) [inline docs/conf.py]: can you remove the merge conflicts here?

**Review by gr2m — CHANGES_REQUESTED** [review]

File: docs/conf.py
Comment (NimmiW) [reply docs/conf.py]: ok @gr2m I will correct it :)

File: docs/api/client/hoodie.account.rst
Comment (gr2m) [reply docs/api/client/hoodie.account.rst:666]: Good idea! I’d suggest we create a follow up issue for that so that we can merge this in, and then improve it later

Comment (NimmiW) [reply docs/api/client/hoodie.account.rst:666]: I did the changes @gr2m now we can merge it i think :)
@pmbanugo By the way yes me should definitely add the things you mentioned in anither issue

Comment (pmbanugo) [reply docs/api/client/hoodie.account.rst:666]: Yeah agreed. We can always make improvement and adjustment. Well done @NimmiW

**Review by pmbanugo — APPROVED** [review]

**Comment by NimmiW** [comment]
Thank you @pmbanugo :) @gr2m or @pmbanugo can anyone of you create a new issue for the improvements of client API with details?

**Comment by NimmiW** [comment]
Hi @gr2m @fozy81 @pmbanugo shall we merge this to master branch if all the things up to now are okay with API doc? :D

**Review by gr2m — APPROVED** [review]

**Comment by gr2m** [comment]
great work @NimmiW 👏

**Pull request merged**
